src/words/environment/WordsObject.java

```diff
@@ -216,11 +223,21 @@ public void executeNextAction(Environment environment) throws WordsProgramExcept
 				
 				Action action = actionQueue.pop();
 				lastAction = action;
+				lastCell = new Position(cell);
 				action.execute(this, environment);
+				if (!lastCell.equals(cell)) {
+					this.movedInTurn = true;
+				}
 			} else {
 				lastAction = new WaitAction(environment.getCurrentScope());
 			}
 		}
+		// Store the last position based on actions
+		lastCell = new Position(cell);
```

Comment:
On the surface, this line is weird to me.  It sets lastCell to be cell, which means they are the same most of the time (except for a short period inside this method).  What's the reason it's here?